### ISSUE TITLE
Assertion "0 <= p4 && p4 <= 1" failed.

### ISSUE THREAD

**Issue by horenbergerb** [horenbergerb/BernoulliDiffusion#22]
Sometimes the program crashes with something like this (I used CUDA_LAUNCH_BLOCKING to get an accurate traceback):

```
/opt/conda/conda-bld/pytorch_1659484809662/work/aten/src/ATen/native/cuda/DistributionTemplates.h:593: operator(): block: [2,0,0], thread: [447,0,0] Assertion `0 <= p4 && p4 <= 1` failed.
Traceback (most recent call last):
  File "/home/captdishwasher/horenbergerb/BernoulliDiffusion/train.py", line 49, in <module>
    diffusion_model = trainer.train()
  File "/home/captdishwasher/horenbergerb/BernoulliDiffusion/BernoulliDiffusion/trainer.py", line 125, in train
    self.log_data_()
  File "/home/captdishwasher/horenbergerb/BernoulliDiffusion/BernoulliDiffusion/trainer.py", line 84, in log_data_
    post_sample = self.diffusion_model.p_sample(self.cfg.num_examples, self.example_seed)
  File "/home/captdishwasher/miniconda3/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/captdishwasher/horenbergerb/BernoulliDiffusion/BernoulliDiffusion/diffusion_model.py", line 51, in p_sample
    x = torch.bernoulli(self.p_conditional_prob(x, cur_t))
RuntimeError: CUDA error: device-side assert triggered
```

This would happen if `torch.bernoulli` was getting passed values outside [0,1] by `self.model(x_t, t)`. In turn, this might be feasible if you're getting NaNs or something.

The issue might start with math_utils producing values which are too small/large, causing NaNs to get into the system.

Not sure how to fix this, but possibly the issue tends to happen with large batch sizes?

**Comment by horenbergerb** [comment]
It seems like the root cause of this issue is the learning rate being too large. Presumably it's causing the model weights to go bonkers (despite my gradient clipping...)
The best thing to do would probably be to add an exception when NaN loss occurs and warn that this frequently occurs due to learning rate being too high.

**Comment by horenbergerb** [comment]
Issue should be addressed in #29
